### PR TITLE
[G]VIF

### LINKED ISSUE #428: Add VIF?
There was a request for VIF on [Discourse](https://discourse.julialang.org/t/regression-variance-inflation-factor-vif/60233/5). 
It's pretty simple to add using the fact that VIFs = the diagonal elements of the inverse of the correlation matrix.
-Belsley, D. A., E. Kuh, and R. E. Welsch. Regression Diagnostics. Hoboken, NJ: John Wiley & Sons, 1980.

```julia
vif_GLM(mod) = diag(inv(cor(mod.model.pp.X[:,2:end])))

using DataFrames, GLM, Statistics, LinearAlgebra, RDatasets 
airquality = rename(dataset("datasets", "airquality"), "Solar.R" => "Solar_R")
#
y =  airquality.Wind
X1 = airquality.Temp    
X2 = airquality.Solar_R  
X3 = airquality.Ozone 
d = DataFrame(X1=X1, X2=X2, X3=X3, y=y)
d = d[completecases(d), :]
X = [one.(d.X1) d.X1 d.X2 d.X3]  .|> Float64
y = d.y .|> Float64
#
β_LS = X \ y
vifm = diag(inv(cor(X[:,2:end])))
#
m_y  = lm(@formula(y ~ 1 + X1 + X2 + X3), d)
m_X1 = lm(@formula(X1 ~ 1 + X2 + X3), d)
m_X2 = lm(@formula(X2 ~ 1 + X1 + X3), d)
m_X3 = lm(@formula(X3 ~ 1 + X1 + X2), d)

vif_X1 = 1.0/(1.0-r2(m_X1))
vif_X2 = 1.0/(1.0-r2(m_X2))
vif_X3 = 1.0/(1.0-r2(m_X3))
#
vifm ≈ [vif_X1; vif_X2; vif_X3;]
#
vif_GLM(m_y) ≈ vifm ≈ [vif_X1; vif_X2; vif_X3;]
```

or
```julia
function vif_glm1(mod::StatsModels.TableRegressionModel)
    diag(inv(cor(mod.model.pp.X[:,2:end])))
end
```

### PR DESCRIPTION
**Pull request by palday**
closes #428

### CODE REVIEW

File: src/linpred.jl [hunk 1]
```diff
@@ -362,7 +362,7 @@ fitted(m::LinPredModel) = m.rr.mu
 predict(mm::LinPredModel) = fitted(mm)
 residuals(obj::LinPredModel) = residuals(obj.rr)
 
-function formula(obj::LinPredModel)
+function StatsModels.formula(obj::LinPredModel)
```
Comment (bkamins) [inline src/linpred.jl:365]: While we are at it. When is it called. When I do:
```
julia> formula(lm(x, y))
ERROR: type LinearModel has no field fr

julia> formula(glm(x, y, Normal()))
ERROR: type GeneralizedLinearModel has no field fr
```
other methods are called.

Do we have tests for different cases when formula is not present?

File: test/runtests.jl [hunk 2]
```diff
@@ -2011,3 +2011,12 @@ end
     @test_throws ArgumentError lm(@formula(OptDen ~ Carb), form; method=:pr)
     @test_throws ArgumentError glm(@formula(OptDen ~ Carb), form, Normal(); method=:pr)
 end
+
+@testset "[G]VIF" begin
+    duncan = RDatasets.dataset("car", "Duncan")
+    lm1 = lm(@formula(Prestige ~ 1 + Income + Education), duncan)
+    @test termnames(lm1)[2] == coefnames(lm1)
```
Comment (bkamins) [inline test/runtests.jl:2018]: 1. do we have tests when `coefnames` and `termnames` differ?
2. do we have a decision what should be done in the case of `lm(X, y)` (i.e. model fitted without formula, it still prints variable names as `x1` etc.)

File: test/runtests.jl [hunk 3]
```diff
@@ -2011,3 +2011,12 @@ end
...
+    @test vif(lm1) ≈ gvif(lm1)
+    lm2 = lm(@formula(Prestige ~ 1 + Income + Education + Type), duncan)
+    @test gvif(lm2; scale=true) ≈ [1.486330, 2.301648, 1.502666] atol=1e-4
```
Comment (bkamins) [inline test/runtests.jl:2021]: 1. Can you please add a comment on where these values are taken from?
2. Also do we have tests for `vif`/`gvif` for `glm`?
3. Do we have tests for `vif`/`gvif` for models without formula?
4. Do we have tests for `vif`/`gvif` for models that have complex formulas, something like e.g `@formula(y~(1+a*(b+log(c)))&(1+d))`? (of course this is artificial, but I hope it is clear what I mean

File: test/runtests.jl (hunk 2 above)
Comment (palday) [reply test/runtests.jl:2018]: 1. This falls back to StatsModels -- the test there is just making sure we've successfully imported and exported the symbol.
2. on `master`,`termnames` will error based on there being no formula (`formula` will return `nothing`).

File: test/runtests.jl (hunk 3 above)
Comment (palday) [reply test/runtests.jl:2021]: These are just the StatsModels tests carried forward to models actually fitted here. :smile: But I can add a cross reference.

File: src/linpred.jl (hunk 1 above)
Comment (palday) [reply src/linpred.jl:365]: Hmmm, will investigate. I thought we caught this when Milan removed TableRegressionModel.

Comment (palday) [reply src/linpred.jl:365]: On current master:

```julia
julia> formula(lm(ones(10, 1),  randn(10)))
ERROR: ArgumentError: model was fitted without a formula
Stacktrace:
 [1] formula(obj::LinearModel{GLM.LmResp{Vector{Float64}}, GLM.DensePredChol{Float64, LinearAlgebra.CholeskyPivoted{Float64, Matrix{Float64}, Vector{Int64}}}})
   @ GLM ~/Code/GLM.jl/src/linpred.jl:366
 [2] top-level scope
   @ REPL[13]:1

julia> formula(glm(ones(10, 1),  randn(10), Normal()))
ERROR: ArgumentError: model was fitted without a formula
Stacktrace:
 [1] formula(obj::GeneralizedLinearModel{GLM.GlmResp{Vector{Float64}, Normal{Float64}, IdentityLink}, GLM.DensePredChol{Float64, LinearAlgebra.CholeskyPivoted{Float64, Matrix{Float64}, Vector{Int64}}}})
   @ GLM ~/Code/GLM.jl/src/linpred.jl:366
 [2] top-level scope
   @ REPL[14]:1
```

(will have to keep this in mind for the backport to 1.x where we still have TableRegressionModel)

File: test/runtests.jl (hunk 2 above)
Comment (palday) [reply test/runtests.jl:2018]: I think `termnames` should not be defined if there is no formula -- there are only Terms when there is a formula.

Comment (bkamins) [reply test/runtests.jl:2031]: So what should a user do to perform VIF analysis for the `model - lm(X, y)` case?

Comment (palday) [reply test/runtests.jl:2031]: `vif` works, but not `gvif`. So I think they can still do `vif`. If they're able to construct a model matrix directly for something with non trivial contrast coding, then they could probably also do adapt the `gvif` source to extract the correct columns.

**Review by bkamins — APPROVED** [review]

File: src/GLM.jl [hunk 4]
```diff
@@ -21,7 +22,7 @@ module GLM
     export coef, coeftable, confint, deviance, nulldeviance, dof, dof_residual,
            loglikelihood, nullloglikelihood, nobs, stderror, vcov, residuals, predict,
            fitted, fit, fit!, model_response, response, modelmatrix, r2, r², adjr2, adjr²,
-           cooksdistance, hasintercept, dispersion
+           cooksdistance, hasintercept, dispersion, vif, gvif, termnames
```
[the remark this answers is not in the document]
Comment (nalimilan) [reply src/GLM.jl:25]: Yeah but there shouldn't be breaking changes in StatsModels minor releases, and anyway users who need these functions will do `using StatsModels`.

**Pull request merged**